File: Editorial/85_Maximal_Rectangle/C/3_Using_Histograms_Stack.c
int largestRectangleArea(int* heights, int heightsSize) {
    int stack[heightsSize];
    int top = -1;
    int maxarea = 0;
    for (int i = 0; i <= heightsSize; ++i) {
        while (top != -1 &&
               (i == heightsSize || heights[stack[top]] >= heights[i])) {
            int height = heights[stack[top--]];
            int width = (top == -1) ? i : i - stack[top] - 1;
            int area = height * width;
            if (area > maxarea) maxarea = area;
        }
        stack[++top] = i;
    }
    return maxarea;
}
int maximalRectangle(char** matrix, int matrixSize, int* matrixColSize) {
    int maxarea = 0;
    if (matrixSize == 0) return 0;
    int dp[matrixColSize[0]];
    memset(dp, 0, sizeof(dp));
    for (int i = 0; i < matrixSize; i++) {
        for (int j = 0; j < matrixColSize[0]; j++) {
            dp[j] = matrix[i][j] == '1' ? dp[j] + 1 : 0;
        }
        maxarea = fmax(maxarea, largestRectangleArea(dp, matrixColSize[0]));
    }
    return maxarea;
}

File: Editorial/0000-0100/0085_Maximal_Rectangle/Go/3_Using_Histograms_Stack.go
func maximalRectangle(matrix [][]byte) int {
    maxArea := 0
    width := len(matrix[0])
    heights := make([]int, width)
    for i := 0; i < len(matrix); i++ {
        for j := 0; j < len(matrix[i]); j++ {
            if matrix[i][j] == '1' {
                heights[j]++
            } else {
                heights[j] = 0
            }
        }
        maxArea = int(math.Max(float64(maxArea), float64(leetcode84(heights))))
    }
    return maxArea
}

func leetcode84(height []int) float64 {
    maxArea := 0.0
    stack := make([]int, 0)
    stack = append(stack, -1)
    for i := 0; i < len(height); i++ {
        for stack[len(stack)-1] != -1 && height[i] <= height[stack[len(stack)-1]] {
            pop := stack[len(stack)-1]
            stack = stack[:len(stack)-1]
            maxArea = math.Max(
                maxArea,
                float64(height[pop]*(i-stack[len(stack)-1]-1)),
            )
        }
        stack = append(stack, i)
    }
    for stack[len(stack)-1] != -1 {
        pop := stack[len(stack)-1]
        stack = stack[:len(stack)-1]
        maxArea = math.Max(
            maxArea,
            float64(height[pop]*(len(height)-stack[len(stack)-1]-1)),
        )
    }
    return maxArea
}
